### PR TITLE
[ysh language] and / or operators do implicit bool(x) on operands

### PR DESCRIPTION
**Pull request by andychu**
Test that it matches Python.

This is #1627

Related discussion - https://oilshell.zulipchat.com/#narrow/stream/384942-language-design/topic/Bool.28x.29.2C.20'foo'.20or.20'bar'

### CODE REVIEW

**Comment by andychu** [comment]
Someone already noticed recent changes!  (that's good)

It seems that Python and JavaScript agree on all this behavior, so I think we need it for YSH

Comments welcome

A pretty common idiom is `var x = param or 'default'`

File: ysh/expr_eval.py [hunk 1]
```diff
@@ -408,6 +408,59 @@ def AsPosixEre(eggex):
...
+      return len(val.s) != 0
+
+    # OLD TYPES
+    elif case(value_e.MaybeStrArray):
+      val = cast(value.MaybeStrArray, UP_val)
+      return len(val.strs) != 0
+
+    elif case(value_e.MaybeStrArray):
+      val = cast(value.MaybeStrArray, UP_val)
+      return len(val.strs) != 0
```
Comment (melvinw) [inline ysh/expr_eval.py:433]: was one of these supposed to be `AssocArray`?

**Review by melvinw — COMMENTED** [review]
Thanks! Just one thing inline

**Comment by andychu** [comment]
Ahh great catch, thanks!

**Comment by andychu** [comment]
merged separately

**Pull request closed without merging**